File: leetcode/subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        length = len(nums)

        def backtrack(nums: List[int], l: int, cur: List[int]):
            res.append(cur.copy())

            for i in range(l, length):
                # take nums[i]
                cur.append(nums[i])

                backtrack(nums, i + 1, cur)

                # leave nums[i]
                cur.pop()

        backtrack(nums, 0, [])

        return res

File: leetcode/subsets.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<vector<int>> subsets(vector<int>& nums)
    {
        vector<vector<int>> res;
        vector<int> cur;

        auto length = nums.size();

        function<void(int, vector<int>&)> backtrack = [&](int l, vector<int> cur) {
            res.push_back(cur);

            if (l >= length)
                return;

            for (auto i = l; i < length; ++i) {
                // take i
                cur.push_back(nums[i]);

                backtrack(i + 1, cur);

                cur.pop_back();
            }
        };

        backtrack(0, cur);

        return res;
    }
};
